extend batch_callback for options (#1525)
yewstack/yew: docs/concepts/components/callbacks.md

title: Callbacks
description: ComponentLink and Callbacks
---

The component "link" is the mechanism through which components are able to create callbacks and update themselves.

## ComponentLink API

### send_message

Sends a message to the component.
Messages are handled by the `update` method which decides whether the component should re-render.

### send_message_batch

Sends multiple messages to the component at the same time.
This is similar to `send_message` but if any of the messages cause the `update` method to return `true`,
the component will re-render after all messages in the batch have been processed.

If the given vector is empty, this function doesn't do anything.

### callback

Create a callback that will send a message to the component when it is executed.

> Under the hood, it will call `send_message` with the message that is returned by the provided closure.

Reviewer: not sure if "that is" is necessary
```suggestion
Under the hood, it will call `send_message` with the message returned by the provided closure.
```

> This method has a version that accepts an `FnOnce` instead, `callback_once`.

Reviewer: Should this say something like (I feel it's not 100% clear about the name of the name of the alternative method `callback_once`):
```suggestion
There is a different method (called `callback_once`) which accepts a `FnOnce` instead of a `Fn`.
``` 
?

Reviewer 2: I agree that this improves the flow but can we drop the parentheses?

Reviewer: Yup.

Reviewer: (n.b. I have updated the original suggestion)

// It can be simplified it to this:
let cb = link.callback(Msg::Text);

// Will send `Msg::Text("Hello World!")` to the component.
cb.emit("Hello World!".to_owned());
```

### batch_callback

Create a callback that will send a batch of messages to the component when it is executed.
The difference to `callback` is that the closure passed to this method doesn't have to return a message.
Instead, the closure can return either `Vec<Msg>` or `Option<Msg>` where `Msg` is the component's message type.

`Vec<Msg>` is treated as a batch of messages and uses `send_message_batch` under the hood.

`Option<Msg>` calls `send_message` if it is `Some`. If the value is `None`, nothing happens.
This can be used in cases where, depending on the situation, an update isn't required.

This is achieved using the `SendAsMessage` trait which is only implemented for these types.
You can implement `SendAsMessage` for your own types which allows you to use them in `batch_callback`.

Like `callback`, this method also has a `FnOnce` counterpart, `batch_callback_once`.
The same restrictions apply as for `callback_once`.

## Callbacks

_\(This might need its own short page.\)_

> Callbacks are used to communicate with services, agents, and parent components within Yew. They are just an `Fn` wrapped by an `Rc` to allow them to be cloned.

Reviewer: I'm not sure if this is the correct form of the article for `Fn` (`Fn` is short "function" which would be "a function" rather than "an function"). 🤷🏼 
```suggestion
Callbacks are used to communicate with services, agents, and parent components within Yew. They are just an `Fn` wrapped by an `Rc` to allow them to be cloned.
```

Reviewer 2: It's always kind of jarring in these situations. "Fn" might be short for function but when I see it like that I read it as "eff-enn" (not exactly IPA but it will have to do :D) which most definitely starts with a vowel sound.
Perhaps we should try to avoid adding articles to types?

```suggestion
Callbacks are used to communicate with services, agents, and parent components within Yew.
Internally their type is just `Fn` wrapped in `Rc` to allow them to be cloned.
```

Reviewer: Yeah sounds good. I'm not particularly fussed about the rules we have so long as we apply the rules we set consistently. 😁

They have an `emit` function that takes their `<IN>` type as an argument and converts that to a message expected by its destination. If a callback from a parent is provided in props to a child component, the child can call `emit` on the callback in its `update` lifecycle hook to send a message back to its parent. Closures or Functions provided as props inside the `html!` macro are automatically converted to Callbacks.

> Simple usage of callbacks can look something like this:

Reviewer: ```suggestion
A simple use of a callback might look something like this:
```

> Simple usage of callbacks can look something like this:

Reviewer: ```suggestion
A simple use of a callback might look something like this:
```

```rust
let onclick = self.link.callback(|_| Msg::Clicked);
html! {
    <button onclick=onclick>{ "Click" }</button>
}
```

> The callback is always given some value. For example, the `onclick` handler will pass a `MouseEvent`. The handler can then decide what kind of message the event should resolve to. This message is scheduled for the next update loop unconditionally.

Reviewer: ```suggestion
The function passed to `callback` must always take a parameter. For example, the `onclick` handler requires a function which takes a parameter of type `MouseEvent`. The handler can then decide what kind of message the should be sent to the component. This message is scheduled for the next update loop unconditionally.
```

Reviewer 2: typo: "what kind of message **the** should be"

Reviewer: Thanks for the catch 😄!
